### ISSUE TITLE
Transform is occasionally applied without pressing button

### ISSUE THREAD

**Issue by Eldresh** [josdejong/svelte-jsoneditor#401]
I cannot begin to tell you how much I wish I could consistently reproduce this error, but its reasonably consistent when you make an error typing in a function. A good example (in my experience) is typing `.splice(2)` and then deleting it, which will leave you with 2 things in your array or an empty array without ever once hitting 'transform' and without the ability to undo it, resulting in a loss of data.

I am using the website, [https://jsoneditoronline.org/](url), and encounter this unfortunately quite often.

**Browser:** Chrome 120.0.6099.217 (Official Build) (64-bit)

**Comment by josdejong** [comment]
Thanks for reporting, I can consistently reproduce the issue the way you describe it, like:

- Open the JSON Editor in tree mode
- Open a JSON array with say 10 items
- Open the Transform modal
- Type something like the following as query:
   ```ks
    function query (data) {
      return data.splice(2)
    }
    ``` 
- Close the transform modal without applying the query (press ESC, or click the X button top right)
- Click "Collapse All", followed by "Expand All", to see that now only 2 items are left in the array, whereas the data should be unchanged

The reason is that the `splice` method mutates the original array. In the query the `data` variable points to the real data, and if you alter it, it is changed for real. The same happens for example if you do say `array[0] = null` inside the query, tha will clear the first item.

To solve this, we'll have to make sure you can't accidentally mutate the oringial `data`. Maybe we can use [Mutative](https://mutative.js.org/docs/intro) there somehow. Or maybe I can wrap the `data` in some proxy, let me give that a try.

**Comment by josdejong** [comment]
This looks like a promising approach: https://github.com/kourge/readonly-proxy

**Comment by josdejong** [comment]
I tried out `readonly-proxy` but ran into some issues. Solved now via 337f812 (not yet published)

**Comment by josdejong** [comment]
Fixed now in `svelte-jsoneditor@0.21.5` and JSON Editor Online `v6.15.1`. Can you give it a try @Eldresh?

**Comment by Eldresh** [comment]
> Fixed now in `svelte-jsoneditor@0.21.5` and JSON Editor Online `v6.15.1`. Can you give it a try @Eldresh?

Yup. Now it gives me TypeError: unable to delete property, which is much better that accidentally losing everything.

**Comment by josdejong** [comment]
👍
